rust/windows-client/src-tauri/src/client/gui.rs

```diff
@@ -403,6 +407,70 @@ impl Controller {
 
         Ok(())
     }
+
+    fn copy_resource(&self, id: &str) -> Result<()> {
+        let Some(session) = &self.session else {
+            bail!("app is signed out");
+        };
+        let resources = session.callback_handler.resources.load();
+        let id = ResourceId::from_str(id)?;
+        let Some(res) = resources.iter().find(|r| r.id() == id) else {
+            bail!("resource ID is not in the list");
+        };
+        let mut clipboard = arboard::Clipboard::new()?;
+        // TODO: Make this a method on `ResourceDescription`
+        match res {
+            ResourceDescription::Dns(x) => clipboard.set_text(&x.address)?,
+            ResourceDescription::Cidr(x) => clipboard.set_text(&x.address.to_string())?,
+        }
+        Ok(())
+    }
+
+    async fn handle_deep_link(&mut self, url: &url::Url) -> Result<()> {
+        let Some(auth) = client::deep_link::parse_auth_callback(url) else {
+            bail!("couldn't parse scheme request");
+        };
+
+        let token = auth.token.clone();
+        // spawn_blocking because keyring access is I/O
+        if let Err(e) = spawn_blocking(move || {
+            let entry = keyring_entry()?;
+            entry.set_password(token.expose_secret())?;
+            Ok::<_, anyhow::Error>(())
+        })
+        .await?
```

Comment:
you could use double `??` here to have consistent error handling

Reply:
Yeah if this is `anyhow` already then why didn't I use `context`? I'll make a note to fix it some time